pythoncode/usbTrainer.py

```diff
@@ -92,477 +115,1090 @@
+
+                if Channel == ant.channel_VHU_s and DeviceTypeID == ant.DeviceTypeID_VHU:
+                    dataHandled = True
+                    self.__AntVHUpaired    = True
+                    self.__DeviceNumberVHU = DeviceNumber
+
+            #-------------------------------------------------------------------
+            # Outer loop does not need to handle channel_VHU_s messages
+            #-------------------------------------------------------------------
+            dataHandled = True
 
         #-----------------------------------------------------------------------
-        # Set configuration
+        # Send messages, leave receiving to the outer loop
         #-----------------------------------------------------------------------
-        if dev != False:
-            dev.set_configuration()
-            if trainer_type == tt_iMagic:   dev.set_interface_altsetting(0, 1)
+        if messages:
+            self.AntDevice.Write(messages, False, False)
 
-        #-----------------------------------------------------------------------
-        # InitialiseTrainer (will not read cadence until initialisation byte is sent)
-        #-----------------------------------------------------------------------
-        if dev != False and LegacyProtocol == False:
-            data = struct.pack (sc.unsigned_int, 0x00000002)
-            if debug.on(debug.Data2):
-                logfile.Write ("InitialiseTrainer data=%s (len=%s)" % (logfile.HexSpace(data), len(data)))
-            dev.write(0x02,data)
+        return dataHandled
+
+#-------------------------------------------------------------------------------
+# c l s T a c x U s b T r a i n e r
+#-------------------------------------------------------------------------------
+class clsTacxUsbTrainer(clsTacxTrainer):
+    #---------------------------------------------------------------------------
+    # Convert WheelSpeed --> Speed in km/hr
+    # SpeedScale must be defined in sub-class
+    #---------------------------------------------------------------------------
+    # WheelSpeed = 1 mm/sec <==> 1.000.000 / 3600 km/hr = 1/278 km/hr
+    # TotalReverse: Other implementations are using values between 1/277 and 1/360.
+    # A practical test shows that Cadence=92 gives 40km/hr at a given ratio
+    #		factor 289.75 gives a slightly higher speed
+    #		factor 301 would be good (8-11-2019)
+    #---------------------------------------------------------------------------
+    def Wheel2Speed(self):
+        self.SpeedKmh = round(self.WheelSpeed / self.SpeedScale, 1)
+
+    def Speed2Wheel(self, SpeedKmh):
+        return int(SpeedKmh * self.SpeedScale)
+    
+    #---------------------------------------------------------------------------
+    # Refresh()
+    #---------------------------------------------------------------------------
+    def Refresh(self, QuarterSecond, TacxMode):
+        super().Refresh(QuarterSecond, TacxMode)

```

Comment:
https://en.wikipedia.org/wiki/Call_super

There are several places still where I firmly believe that more refactoring will yield cleaner, less coupled code. It's often the case that one change enables the next, or that one smell masks another, and personally, I never stop looking at code and asking how might this be better expressed. There's a few things that still stand out to me.

That's not to diminish any of the work you've done, and I applaud your efforts on this software and the community of users it has enabled to get riding with modern software, hopefully more to come. And for you being new to python and OO, etc.

I accept some items are down to style and we've discussed them already, though I still believe in "decouple, decouple, decouple". I'm not implying that I think you should refactor anything else now. I am not intending to raise any more of these thoughts uninvited.

The user of call-super just made me chuckle and I thought I'd just post a wee kind of refactoring-review-summary type of thing as we approach the time to merge this PR. A combination of a well-done to you and some enthusiastic support from me.

Reply:
A lot of text (too much?) to express a concern.
The implementation is: do the standard class-thing AND do also the ivortex-key conversion.
I think it's coded as python suggests.
Alternative is to create _ReceiveFromTrainer() for ivortex, semantically probably a more elegant implementation

Reply:
A lot of text (too much?) to express a concern.
The implementation is: do the standard class-thing AND do also the ivortex-key conversion.
I think it's coded as python suggests.
Alternative is to create _ReceiveFromTrainer() for ivortex, semantically probably a more elegant implementation

Reply:
As it is, the specific code calls back to the common code, and must do so in each child class for correct behaviour. It's functionally correct, a natural way to think of it, but implemented this way round has the name of "call-super" and there is a cleaner way. 

To refactor and avoid use of call-super is to have the parent/common function not be overridden and for it to call a differently named and indeed overridden function on the child class - ie the common code calls out to the specific code. I used "Impl" as a suffix for such functions in my branch.


As the extra text suggested, I think you're close to wanting to release so while I enjoy refactor and review, and would like to see/do more, I mention it and move on.

Reply:
...requiring the parent to be changed (add additional call) which basically is against inheritance.
One could standardize that each call has a ..before() and after() but that would complicated things as well.

I adhere to the idea of clean code. Question here is: what would the correct code be, given that the existing classes are frozen and only clsVortex may be changed?

Reply:
And yes - I want to release 3.0; not excluding further devs. People are waiting for it!

Reply:
Leave as is. Functionally, it doesn't matter. Also maintenance impact is minimal.